scipy/sparse/linalg/_dsolve/_superlumodule.c

```diff
@@ -263,6 +264,152 @@ static PyObject *Py_gstrf(PyObject * self, PyObject * args,
+  if (L_type != U_type) {
+      PyErr_SetString(PyExc_TypeError,
+                      "nzvals types of L and U differ");
+      return NULL;
+  }
+  if (!CHECK_SLU_TYPE(L_type)) {
+      PyErr_SetString(PyExc_TypeError,
+                      "nzvals is not of a type supported by SuperLU");
+      return NULL;
+  }
+
+  /* Create SuperLU matrices out of L and U. */
+  int* L_col_to_sup = intMalloc(L_N+1), *L_sup_to_col = intMalloc(L_N+1);
+  for(int i=0; i<=L_N; i++){
+    L_col_to_sup[i] = i;
+    L_sup_to_col[i] = i;
+  }
+  L_col_to_sup[L_N] = L_N-1;
+  SuperMatrix L_super = {0}, U_super={0};
+  int L_conv_err = SparseFormat_from_spMatrix(&L_super, L_N, L_N, L_nnz, -1,
+    (PyArrayObject*)L_nzvals, (PyArrayObject*)L_colind, (PyArrayObject*)L_rowptr, L_type, SLU_SC, SLU_TRLU, L_col_to_sup,L_sup_to_col);
+  if(L_conv_err){
+    return NULL;
+  }
+  int U_conv_err = SparseFormat_from_spMatrix(&U_super, U_N, U_N, U_nnz, 0,
+    (PyArrayObject*)U_nzvals, (PyArrayObject*)U_rowind, (PyArrayObject*)U_colptr, U_type, SLU_NC, SLU_TRU, NULL,NULL);
+  if(U_conv_err){
+    Destroy_SuperMatrix_Store((SuperMatrix*)&L_super);
+    return NULL;
+  }
+
+  /* Read right-hand-side/solution vector. */
+  PyArrayObject* X_arr = (PyArrayObject*)PyArray_FROMANY(
+      (PyObject*)X_py, L_type, 1, 2,
+      NPY_ARRAY_F_CONTIGUOUS | NPY_ARRAY_ENSURECOPY);
+  if (X_arr == NULL){
+      SUPERLU_FREE((void*)L_col_to_sup); SUPERLU_FREE((void*)L_sup_to_col);
+      Destroy_SuperMatrix_Store((SuperMatrix*)&L_super);
+      Destroy_SuperMatrix_Store((SuperMatrix*)&U_super);
+      return NULL;
+  }
+  if (PyArray_DIM((PyArrayObject*)X_arr, 0) != L_N) {
+      PyErr_SetString(PyExc_ValueError,
+                      "right hand side array has invalid shape");
+      SUPERLU_FREE((void*)L_col_to_sup); SUPERLU_FREE((void*)L_sup_to_col);
+      Destroy_SuperMatrix_Store((SuperMatrix*)&L_super);
+      Destroy_SuperMatrix_Store((SuperMatrix*)&U_super);
+      Py_DECREF(X_arr);
+      return NULL;
+  }
+
+  SuperMatrix X;
+  if (DenseSuper_from_Numeric((SuperMatrix*)&X, (PyObject*)X_arr)) {
+      SUPERLU_FREE((void*)L_col_to_sup); SUPERLU_FREE((void*)L_sup_to_col);
+      Destroy_SuperMatrix_Store((SuperMatrix*)&L_super);
+      Destroy_SuperMatrix_Store((SuperMatrix*)&U_super);
+      Py_DECREF(X_arr);
+      return NULL;
+  } /* X and X_arr share the same data but X_arr "owns" it. */
+
+  /* Call SuperLU functions. */
+  volatile jmp_buf* jmpbuf_ptr = (volatile jmp_buf *)superlu_python_jmpbuf();
+  int info=0;
+  SuperLUStat_t stat = { 0 };
+  StatInit((SuperLUStat_t *)&stat);
+  int* perm_c = intMalloc(L_N);
+  for(int i=0; i<L_N; i++){
+    perm_c[i] = i;
+  }
+  int* perm_r = perm_c;
+  SLU_BEGIN_THREADS_DEF;
+  SLU_BEGIN_THREADS;
+  if (setjmp(*(jmp_buf*)jmpbuf_ptr)) {
+      SLU_END_THREADS;
+      goto fail;
+  }
+  gstrs(L_type,
+        NOTRANS, &L_super, &U_super, perm_c, perm_r,
+        (SuperMatrix *)&X, (SuperLUStat_t *)&stat, (int *)&info);
+  SLU_END_THREADS;
```

Comment:
Does this use multithreading by default and how does it relate to other usages of SuperLU in sparse?

Reply:
I reordered the statements a bit in commit be7af67d so `SLU_BEGIN_THREADS_DEF`, `SLU_BEGIN_THREADS`, `SLU_END_THREADS` are now used in precisely the same way as in [`Py_gssv`](https://github.com/scipy/scipy/blob/v1.10.0/scipy/sparse/linalg/_dsolve/_superlumodule.c#L162-L168), [`SuperLU_solve`](https://github.com/scipy/scipy/blob/dde50595862a4f9cede24b5d1c86935c30f1f88a/scipy/sparse/linalg/_dsolve/_superluobject.c#L79), [`newSuperLUObject`](https://github.com/scipy/scipy/blob/dde50595862a4f9cede24b5d1c86935c30f1f88a/scipy/sparse/linalg/_dsolve/_superluobject.c#L746). 

My non-expert understanding of the threading aspects is as follows:

The three macros are defined in [`_superluobject.h`](https://github.com/scipy/scipy/blob/v1.10.0/scipy/sparse/linalg/_dsolve/_superluobject.h#L81-L83): Basically, they expand to a declaration of a `PyThreadState*` and calls to `PyEval_SaveThread` and `PyEval_RestoreThread`, respectively. Comparing that with [this part of the Python C API documentation](https://docs.python.org/3/c-api/init.html#releasing-the-gil-from-extension-code), I guess that the purpose of these macros is to release the global interpreter lock (GIL) while SuperLU code is running. SuperLU does not access Python objects, so it does not need to hold the GIL. By releasing the GIL, other  threads may meanwhile acquire the GIL and access Python objects. 

SuperLU is sequential (opposed to SuperLU_MT and SuperLU_DIST, but SciPy uses SuperLU without suffix), except that it may call multi-threaded BLAS routines, according to the [users' guide](https://portal.nersc.gov/project/sparse/superlu/ug.pdf), section 1.4.2. When I run the [reproducing code example](https://github.com/scipy/scipy/issues/16131#issue-1228053542), both `spsolve_triangular` and `spsolve` use only one core on my system.

Reply:
From my novice understanding this seems resolved then, thanks for the deep dive.